### PR TITLE
Feature/6435 rat 3rd party solution integration

### PR DESCRIPTION
**Pull request by KaiTeuber**
## Description
Calling the URL https://s.coronawarn.app on the Smartphone opens the CWA and parse the information provided in the QR-Code:
* failure will show an alert
* success will open exposure submission flow and skip internal QR Code scanner

## Link to Jira
https://jira-ibs.wbs.net.sap/browse/EXPOSUREAPP-6464
https://jira-ibs.wbs.net.sap/browse/EXPOSUREAPP-6465

## Screenshots
Simulator Screen Shot - iPhone 12 Pro - 2021-04-15 at 17 42 33
Simulator Screen Shot - iPhone 12 Pro - 2021-04-15 at 17 42 58

### CODE REVIEW

**Review by janetback — APPROVED** [review]
UA reviewed

**Comment by BLUM78691** [comment]
Du sitzt ja immer noch am Rechner …

File: src/xcode/ENA/ENA/Source/RootCoordinator.swift [hunk 1]
```diff
@@ -52,11 +52,12 @@ class RootCoordinator: RequiresAppDependencies {
...
 	}()
 
-	func showHome(enStateHandler: ENStateHandler) {
-		viewController.clearChildViewController()
-		
-		// Home
-		guard let delegate = delegate else {
+	func showHome(enStateHandler: ENStateHandler, route: Route?) {
+		// only create and init the whole view stack if not done before
+		// there for we check if the homeCoordinator exists
```
Comment (SwissMountainFrog) [inline src/xcode/ENA/ENA/Source/RootCoordinator.swift:57]:
```suggestion
		// therefor we check if the homeCoordinator exists
```

Comment (SwissMountainFrog) [reply src/xcode/ENA/ENA/Source/RootCoordinator.swift:57]: Just a little typo ;-)

File: src/xcode/ENA/ENA/Source/Scenes/ExposureSubmission/ExposureSubmissionCoordinator.swift [hunk 2]
```diff
@@ -44,24 +44,38 @@ class ExposureSubmissionCoordinator: NSObject, RequiresAppDependencies {
...
 
-		/// The navigation controller keeps a strong reference to the coordinator. The coordinator only reaches reference count 0
-		/// when UIKit dismisses the navigationController.
-		let exposureSubmissionNavigationController = ExposureSubmissionNavigationController(
-			coordinator: self,
-			dismissClosure: { [weak self] in
-				self?.dismiss()
-			},
-			rootViewController: initialVC
+	private func showRATInvalidQrCodeAlert() {
```
Comment (SwissMountainFrog) [inline src/xcode/ENA/ENA/Source/Scenes/ExposureSubmission/ExposureSubmissionCoordinator.swift:67]: Maybe a bit niggling, but in the localized strings, we do not camelcase `QR`, in the function name `showRATInvalidQrCodeAlert` we mix both. So might `showRATInvalidQRCode` be an option? There are function name examples in our codebase (eg. `.latestQRCodePosterTemplate()`) where we also do not camelCase it. 
Even later down the code road in this file, there is a function `makeQRInfoScreen`.
Just thinking.

**Review by SwissMountainFrog — CHANGES_REQUESTED** [review]

File: src/xcode/ENA/ENA/Source/Scenes/ExposureSubmission/ExposureSubmissionCoordinator.swift [hunk 3]
```diff
@@ -413,64 +447,77 @@ class ExposureSubmissionCoordinator: NSObject, RequiresAppDependencies {
...
 		)
-		
-		push(topBottomContainerViewController)
+
+		return topBottomContainerViewController
 	}
 
-	private func showQRScreen(isLoading: @escaping (Bool) -> Void) {
-		let scannerViewController = ExposureSubmissionQRScannerViewController(
-			onSuccess: { [weak self] testQrCodeInformation in
```
Comment (naveeddotio) [inline src/xcode/ENA/ENA/Source/Scenes/ExposureSubmission/ExposureSubmissionCoordinator.swift:427]: testQRCodeInformation

File: src/xcode/ENA/ENA/Source/Scenes/Home/HomeCoordinator.swift [hunk 4]
```diff
@@ -88,86 +88,107 @@ class HomeCoordinator: RequiresAppDependencies {
...
-					self?.showSettings(enState: enState)
-				}
-			)
-
-			self.homeState = homeState
-			self.homeController = homeController
-			addToEnStateUpdateList(homeState)
+	private func selectHomeTabSection() {
+		DispatchQueue.main.async { [weak self] in
+			guard let rootViewController = self?.rootViewController,
```
Comment (naveeddotio) [inline src/xcode/ENA/ENA/Source/Scenes/Home/HomeCoordinator.swift:93]: guard let self = self else { return } to remove too many optional stuff?

File: src/xcode/ENA/ENA/Source/AppDelegate & Globals/AppDelegate.swift [hunk 5]
```diff
@@ -161,7 +161,7 @@ class AppDelegate: UIResponder, UIApplicationDelegate, CoronaWarnAppDelegate, Re
 	let taskScheduler: ENATaskScheduler = ENATaskScheduler.shared
 	let contactDiaryStore: DiaryStoringProviding = ContactDiaryStore.make()
 	let eventStore: EventStoringProviding = EventStore.make()
-    let serverEnvironment: ServerEnvironment
+	let serverEnvironment: ServerEnvironment
```
Comment (naveeddotio) [inline src/xcode/ENA/ENA/Source/AppDelegate & Globals/AppDelegate.swift:164]: serverEnvironment - spelling mistake here

File: src/xcode/ENA/ENA/Source/AppDelegate & Globals/AppDelegate.swift [hunk 6]
```diff
@@ -639,15 +637,17 @@ class AppDelegate: UIResponder, UIApplicationDelegate, CoronaWarnAppDelegate, Re
...
+		case .none:
+			coordinator.showHome(enStateHandler: enStateHandler, route: nil)
 
-		if let eventRoute = route {
-			switch eventRoute {
-			case .checkin(let guid):
-				coordinator.showEvent(guid)
-			}
-		}
+		case .checkin(let guid):
```
Comment (naveeddotio) [inline src/xcode/ENA/ENA/Source/AppDelegate & Globals/AppDelegate.swift:644]: checkIn

File: src/xcode/ENA/ENA/Source/Scenes/ExposureSubmission/ExposureSubmissionCoordinator.swift [hunk 7]
```diff
@@ -413,64 +447,77 @@ class ExposureSubmissionCoordinator: NSObject, RequiresAppDependencies {
...
+				self?.registerTestAndGetResult(with: testQrCodeInformation, submissionConsentGiven: true, isLoading: isLoading)
+			}
+		}
+
+		if let testInformation = testInformation {
+			testInformationSuccess(testInformation)
+		} else {
+			let scannerViewController = ExposureSubmissionQRScannerViewController(
+				onSuccess: { [weak self] testQrCodeInformation in
+					self?.presentedViewController?.dismiss(animated: true) {
```
Comment (ArturFriesen) [inline src/xcode/ENA/ENA/Source/Scenes/ExposureSubmission/ExposureSubmissionCoordinator.swift:478]: Should the success part also be dispatched on main?

File: src/xcode/ENA/ENA/Source/Scenes/ExposureSubmission/ExposureSubmissionCoordinator.swift [hunk 8]
```diff
@@ -413,64 +447,77 @@ class ExposureSubmissionCoordinator: NSObject, RequiresAppDependencies {
...
-			}
-		)
+			)
 
-		let qrScannerNavigationController = UINavigationController(rootViewController: scannerViewController)
-		qrScannerNavigationController.modalPresentationStyle = .fullScreen
+			let qrScannerNavigationController = UINavigationController(rootViewController: scannerViewController)
+			qrScannerNavigationController.modalPresentationStyle = .fullScreen
+
+			navigationController?.present(qrScannerNavigationController, animated: true)
```
Comment (ArturFriesen) [inline src/xcode/ENA/ENA/Source/Scenes/ExposureSubmission/ExposureSubmissionCoordinator.swift:517]: Is it Ok, that there can be a call to "self?.showOverrideTestNotice(...)" and after that a call to "navigationController?.present(...)" ?

Seems that there can be 2 navigation calls after each other?

**Review by ArturFriesen — CHANGES_REQUESTED** [review]

**Review by SwissMountainFrog — APPROVED** [review]

**Comment by marcussc** [comment]
Labels please :)

**Review by BLUM78691 — APPROVED** [review]
UA approved

**Review by ArturFriesen — APPROVED** [review]

**Pull request merged**